# Introduction to handling advanced Data Structures with Pandas

## Numpy Library

*Numpy* is a fundamental Python library for scientific computing. Among its features is the creation of multidimensional arrays, which can be treated as vectors, and is very fast when performing mathematical operations on them. Which makes it a necessary tool for projects with high computing and mathematical calculation requirements and that is why it is so popular in the scientific ecosystem.

### Math Operations with Numpy Arrays

To get started, we'll import the library and create our first Numpy Array, then affect it with some basic math. It is common to rename the `Numpy` library as `np`.

In [1]:
# The numpy library is imported. The acronym for the numpy library is usually np.
import numpy as np

# Create an array of numbers
nums = np.array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10])

# print the value of the array
nums

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10])

**It is worth mentioning that a Numpy array is very similar to a native Python list.** The main difference lies in the type of mathematical ordering operations that can be performed on a library array and the native lists of Python. python.

Let's take as a starting point the sum of two `Numpy` arrays, compared to a Python sum of lists:

In [2]:
# Print the sum of the Numpy array
nums+nums

array([ 0,  2,  4,  6,  8, 10, 12, 14, 16, 18, 20])

The size of the list was kept, but its elements were added, and this is the potential and simplicity of Numpy arrays, and because of its vast use in the scientific world.

Let's look at equivalent code with a native Python list:

In [3]:
# Compressed list with addition of its elements
list_nums = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
[x + x for x in list_nums]

[0, 2, 4, 6, 8, 10, 12, 14, 16, 18, 20]

We don't think it's complicated to be creating Python *list comprehensions*, but it would clearly reduce the lines of code in our program using **Numpy** arrays. And it would be impossible to achieve it if we need to add two lists.

*Numpy arrays* can also perform the other operations following the same pattern:

```markdown
<numpy_array> <operation (+, -, *, /)> <numpy_array>
```
*Let's be careful with division if there are zeros in the denominator, remember that it is not possible to divide by zero, in which case an error would appear*

Now, to see the potential of Numpy arrays, what would happen if we tried to add two arrays of size $n$, let's talk arrays of $10,000,000 elements?

In [4]:
#Numpy library is imported
import numpy as np

# Two arrays with 1 million elements are created, where each element has a random value between 0-10
x = np.random.choice(10, 1000000)
y = np.random.choice(10, 1000000)

print(x * y)

[18  0 18 ... 81  0 40]


Did you see it? The multiplication took just milliseconds. And that's where it makes using the library worthwhile.

## Introduction to the Panda Series

Before talking about Dataframes it is important that we do a little review about the **Series**, at the same time we will begin to interact with the Pandas library.

**`Pandas`** is a Python library (like `Numpy`), which is characterized by providing data structures that are fast processing, that are easy to express to make working with relational data it's easy and intuitive. [More information here](https://pandas.pydata.org/docs/getting_started/overview.html)

What is a Serie? How do you create a Serie? Let's see the answers to these questions:

Series are a type of one-dimensional structure *(1D)* very similar to an array, with the characteristic that we can label its data to a certain extent through indices.

To study the potential of the series we are going to interact with the `iris.data` file (./data/iris.data) that comes from the data repository of the [UCI Machine Learning](https://archive-beta.ics.uci.edu/).

In [5]:
import pandas as pd

# The iris.data file is imported and its contents are read
with open("data/iris.data", "r") as DATA:
     iris_dataset = DATA.read()

# The dataset is cleaned and the series is built from the array generated by the `split`
iris_dataset = iris_dataset.split('\n')
iris_dataset = [i.split(',') for i in iris_dataset]

# The series will be created with the dataset data
iris_series = pd.Series(iris_dataset)

# print the string
iris_series

0         [5.1, 3.5, 1.4, 0.2, Iris-setosa]
1         [4.9, 3.0, 1.4, 0.2, Iris-setosa]
2         [4.7, 3.2, 1.3, 0.2, Iris-setosa]
3         [4.6, 3.1, 1.5, 0.2, Iris-setosa]
4         [5.0, 3.6, 1.4, 0.2, Iris-setosa]
                       ...                 
145    [6.7, 3.0, 5.2, 2.3, Iris-virginica]
146    [6.3, 2.5, 5.0, 1.9, Iris-virginica]
147    [6.5, 3.0, 5.2, 2.0, Iris-virginica]
148    [6.2, 3.4, 5.4, 2.3, Iris-virginica]
149    [5.9, 3.0, 5.1, 1.8, Iris-virginica]
Length: 150, dtype: object

*From now on whenever we want to display the content of a series and/or a Dataframe we will use the `head` method, which only displays the first data and not the entire set.*

And that's how easy a basic but powerful data structure is built that gives us the *Pandas* library. Let's explore what we have:

- Each row is a one-dimensional array with five elements, $150$ rows were printed, each with a list of 5 elements, plus the enumeration column.

- The *series* as well as the lists and arrays of *Numpy* have methods and their manipulation is also achieved through indexing.

In [6]:
# Example 1: The first data in the series is printed using the head method.
iris_series.head()

0    [5.1, 3.5, 1.4, 0.2, Iris-setosa]
1    [4.9, 3.0, 1.4, 0.2, Iris-setosa]
2    [4.7, 3.2, 1.3, 0.2, Iris-setosa]
3    [4.6, 3.1, 1.5, 0.2, Iris-setosa]
4    [5.0, 3.6, 1.4, 0.2, Iris-setosa]
dtype: object

Another useful method is the `index` method, it returns the list of indices and their range.

In [7]:
# The array indices are printed, in this case it refers to the first column serie_de_iris.index
iris_series.index

RangeIndex(start=0, stop=150, step=1)

What if we wanted to access the elements of position $120$ and up:

In [8]:
# Print the elements from position 120 onwards
iris_series[120:]

120    [6.9, 3.2, 5.7, 2.3, Iris-virginica]
121    [5.6, 2.8, 4.9, 2.0, Iris-virginica]
122    [7.7, 2.8, 6.7, 2.0, Iris-virginica]
123    [6.3, 2.7, 4.9, 1.8, Iris-virginica]
124    [6.7, 3.3, 5.7, 2.1, Iris-virginica]
125    [7.2, 3.2, 6.0, 1.8, Iris-virginica]
126    [6.2, 2.8, 4.8, 1.8, Iris-virginica]
127    [6.1, 3.0, 4.9, 1.8, Iris-virginica]
128    [6.4, 2.8, 5.6, 2.1, Iris-virginica]
129    [7.2, 3.0, 5.8, 1.6, Iris-virginica]
130    [7.4, 2.8, 6.1, 1.9, Iris-virginica]
131    [7.9, 3.8, 6.4, 2.0, Iris-virginica]
132    [6.4, 2.8, 5.6, 2.2, Iris-virginica]
133    [6.3, 2.8, 5.1, 1.5, Iris-virginica]
134    [6.1, 2.6, 5.6, 1.4, Iris-virginica]
135    [7.7, 3.0, 6.1, 2.3, Iris-virginica]
136    [6.3, 3.4, 5.6, 2.4, Iris-virginica]
137    [6.4, 3.1, 5.5, 1.8, Iris-virginica]
138    [6.0, 3.0, 4.8, 1.8, Iris-virginica]
139    [6.9, 3.1, 5.4, 2.1, Iris-virginica]
140    [6.7, 3.1, 5.6, 2.4, Iris-virginica]
141    [6.9, 3.1, 5.1, 2.3, Iris-virginica]
142    [5.8, 2.7, 5.1, 1.9, Iris

As a *Series* is a one-dimensional *(1D)* array, each row contains all the information returned by the original dataset, represented in a native *Python* array for manipulation. **Technically this makes it difficult** to manipulate the data within each series (for example, finding the mean of the first column, which leads to us wasting the native functionality of a Series, because we would have to create a new one). series for each column of data, then inefficient series.

Let's see an example of how to extract the values of the first column in question, for this, we start loading the original dataset again but this time we only take the first column:

In [9]:
import pandas as pd

with open("data/iris.data", "r") as DATA:
     iris_dataset = DATA.read()

# The dataset is cleaned and the series is built from the array generated by the `split`
iris_dataset = iris_dataset.split('\n')

# The first column is extracted from the dataset and its data is converted to float type
iris_dataset = [float(j[0]) if j[0] != '' else 0 for j in [i.split(',') for i in iris_dataset]]

# The series is created with the dataset data
iris_series_col_1 = pd.Series(iris_dataset)

# The series is printed with the first column of the dataset
iris_series_col_1

0      5.1
1      4.9
2      4.7
3      4.6
4      5.0
      ... 
145    6.7
146    6.3
147    6.5
148    6.2
149    5.9
Length: 150, dtype: float64

And now if we can find some statistics as the mean of all the values, and we'll use *Numpy* for this:

In [10]:
import numpy as np

# The mean of the series is calculated from the mean method of numpy
print('Average data size:')
np.mean(iris_series_col_1)

Average data size:


5.843333333333334

Due to the limitations of the *Series (as we saw previously)* we will have to give way to a new data structure, and this is where the *Dataframes* come into play. But how about we first describe the columns coming from our dataset?

According to the description found on the page we know that it has the following structure:

| No. |      Column          |  Data type    | Possible Values                                   |
|-----|:--------------------:|--------------:|:------------------------------------------------|
| 1   |    sepal length      |   float/cm    | Positives                                       |
| 2   |     sepal width      |   float/cm    | Positives                                       |
| 3   |    petal length      |   float/cm    | Positives                                       |
| 4   |     petal width      |    float/cm   | Positives                                       |
| 5   |        class         |  string/text  | Iris Setosa, Iris Versicolour,  Iris Virginica  |

It is time for us to modify the *Series*, and convert it into a *Dataframe* and use its full potential to our advantage.

## Introduction to DataFrames

A *DataFrame* is an $n$ dimensional array of structured data that can store data of different types. Yes, it's like a spreadsheet or a database table.

The DataFrame are more common than the Series *(They appear everywhere in scientific computing, in the analysis and visualization of data and among others)* and it is for this reason that they will be studied in practice.

Usually you can create a `DataFrame` from different data sources, but this time we will continue using the [`iris`](./data/iris.data) dataset.

In [11]:
import pandas as pd

# We create the DataFrame from iris_dataset, the names parameter allows us to name the columns
df_iris = pd.read_csv("./data/iris.data", names=["long_sepal", "width_sepal", "long_petal", "width_petal", "class_1"])

# Print the first 5 elements of the dataframe
df_iris.head()

long_sepal  width_sepal  long_petal  width_petal      class_1
0         5.1          3.5         1.4          0.2  Iris-setosa
1         4.9          3.0         1.4          0.2  Iris-setosa
2         4.7          3.2         1.3          0.2  Iris-setosa
3         4.6          3.1         1.5          0.2  Iris-setosa
4         5.0          3.6         1.4          0.2  Iris-setosa

*Note: from now on whenever we want to import the content of a local dataset we will use the `read_csv` method of the Pandas library.*

As we saw it was very easy to build the `DataFrame`, now, how about we manipulate it? We could answer questions like: how many kinds of iris do we have? What is the average size of the petal length? all classes or one in particular? Let's see how it's done:

If we want to work with the first column, we can easily do it like this:

In [12]:
df_iris['long_sepal'].head()

0    5.1
1    4.9
2    4.7
3    4.6
4    5.0
Name: long_sepal, dtype: float64

The column elements are stored as a `Range`, which can be easily operated on, for example finding the average sepal length.

In [13]:
### Example 1: Average Sepal Length regardless of class
mean_long_sepal = np.mean(df_iris['long_sepal'])
print('Average Iris Petal Size:')
print(mean_long_sepal)

Average Iris Petal Size:
5.843333333333334


That was easy, find the average of $150$ rows and with it, give a little more meaning to the dataset.

The *Dataframe* have many methods that allow to operate, manipulate, group, filter the data, among others. Let's see some of the most important:

* **groupby**: used to group the rows around the value of one of its columns
* **assign**: used to add new columns from the values of other
* **query**: used to filter the dataset based on some condition *(such as the value of a column)*
* **filter**: filters a Dataframe by columns or rows according to their labels in the indexes. *(Kindly like the `query` method but the filter is based on the indices)
* **value_counts()**: used to count how many values in a column

And now how about we now calculate the number of iris classes we have? Let's see how it's done:

In [14]:
# Example 2: the groupby method is used to group the dataframe by the class column
print('The number of classes in the dataset is: ') 
len(df_iris.groupby('class_1'))

The number of classes in the dataset is: 


3

In [15]:
# how much data there is of each class
df_iris['class_1'].value_counts()

class_1
Iris-setosa        50
Iris-versicolor    50
Iris-virginica     50
Name: count, dtype: int64

Now we are going to give way to the answers of the missing questions, let's see what the mean of the petal length for the Iris-setosa class through the query method:

In [16]:
# Example 2: Average Petal Length for the class Iris-setosa
df_iris_setosa = df_iris.query("class_1 == 'Iris-setosa'")
mean_long_petal = np.mean(df_iris_setosa['long_petal'])
print('Average petal size of the Iris-setosa class:')
print(mean_long_petal)

Average petal size of the Iris-setosa class:
1.464


And now we will use the **Method Chain** to calculate a computed value in a new column using the `assign` method, what does that mean about the Method Chain? on the same object in the same line of code, separating each one by a period `.`
  An example is better than a thousand words:

In [17]:
# Example 3: Computed ratio of petal size for class Iris-virginica
# First the Iris-virginica class elements are filtered, then a new column is created where the width and the length of the petal are divided, finally only the first five elements are shown.
df_iris.query('class_1 == "Iris-virginica"').assign(petal_proportion=lambda i: i['width_petal'] / i['long_petal']).head()

long_sepal  width_sepal  long_petal  width_petal         class_1   
100         6.3          3.3         6.0          2.5  Iris-virginica  \
101         5.8          2.7         5.1          1.9  Iris-virginica   
102         7.1          3.0         5.9          2.1  Iris-virginica   
103         6.3          2.9         5.6          1.8  Iris-virginica   
104         6.5          3.0         5.8          2.2  Iris-virginica   

     petal_proportion  
100          0.416667  
101          0.372549  
102          0.355932  
103          0.321429  
104          0.379310

And with this we managed to answer the questions previously proposed. So it's easy to manipulate a *DataFrame* while playing with its data.

## More about DataFrames


To continue with the practice, it would be good to give a little review on the most basic and standard manipulation ideas of *DataFrames*. We will do this at the same time that we will find the *standard deviation on the length of the petals* of the class *Iris-versicolor*.

This demo, basic manipulation process will be step by step as new columns are added, and for convenience we will work with a portion of the original dataset:

In [18]:
# Example 4: Calculate the standard deviation of the petal length for the Iris Versicolor class
df_iris_versicolor = df_iris.query('class_1 == "Iris-versicolor"').copy()

# Calculate the mean length of the petal
media_long_petal = np.mean(df_iris_versicolor['long_petal'])

# A column is created with the subtraction of the average of each length of the petal
df_iris_versicolor['long_petal_minus_media'] = df_iris_versicolor['long_petal'] - media_long_petal

# Create a column with the difference above the square
df_iris_versicolor['long_petal_minus_media_square'] = (df_iris_versicolor['long_petal'] - media_long_petal)**2

# Calculate the mean squared
media_long_petal_error = np.mean(df_iris_versicolor['long_petal_minus_media_square'])

# Calculate standard deviation
print(f'DevSt = {np.square(media_long_petal_error)}')

# Print the dataset
df_iris_versicolor


DevSt = 0.04682896000000002


long_sepal  width_sepal  long_petal  width_petal          class_1   
50         7.0          3.2         4.7          1.4  Iris-versicolor  \
51         6.4          3.2         4.5          1.5  Iris-versicolor   
52         6.9          3.1         4.9          1.5  Iris-versicolor   
53         5.5          2.3         4.0          1.3  Iris-versicolor   
54         6.5          2.8         4.6          1.5  Iris-versicolor   
55         5.7          2.8         4.5          1.3  Iris-versicolor   
56         6.3          3.3         4.7          1.6  Iris-versicolor   
57         4.9          2.4         3.3          1.0  Iris-versicolor   
58         6.6          2.9         4.6          1.3  Iris-versicolor   
59         5.2          2.7         3.9          1.4  Iris-versicolor   
60         5.0          2.0         3.5          1.0  Iris-versicolor   
61         5.9          3.0         4.2          1.5  Iris-versicolor   
62         6.0          2.2         4.0          1.0  Iris-versicolor   
63         6.1          2.9         4.7          1.4  Iris-versicolor   
64         5.6          2.9         3.6          1.3  Iris-versicolor   
65         6.7          3.1         4.4          1.4  Iris-versicolor   
66         5.6          3.0         4.5          1.5  Iris-versicolor   
67         5.8          2.7         4.1          1.0  Iris-versicolor   
68         6.2          2.2         4.5          1.5  Iris-versicolor   
69         5.6          2.5         3.9          1.1  Iris-versicolor   
70         5.9          3.2         4.8          1.8  Iris-versicolor   
71         6.1          2.8         4.0          1.3  Iris-versicolor   
72         6.3          2.5         4.9          1.5  Iris-versicolor   
73         6.1          2.8         4.7          1.2  Iris-versicolor   
74         6.4          2.9         4.3          1.3  Iris-versicolor   
75         6.6          3.0         4.4          1.4  Iris-versicolor   
76         6.8          2.8         4.8          1.4  Iris-versicolor   
77         6.7          3.0         5.0          1.7  Iris-versicolor   
78         6.0          2.9         4.5          1.5  Iris-versicolor   
79         5.7          2.6         3.5          1.0  Iris-versicolor   
80         5.5          2.4         3.8          1.1  Iris-versicolor   
81         5.5          2.4         3.7          1.0  Iris-versicolor   
82         5.8          2.7         3.9          1.2  Iris-versicolor   
83         6.0          2.7         5.1          1.6  Iris-versicolor   
84         5.4          3.0         4.5          1.5  Iris-versicolor   
85         6.0          3.4         4.5          1.6  Iris-versicolor   
86         6.7          3.1         4.7          1.5  Iris-versicolor   
87         6.3          2.3         4.4          1.3  Iris-versicolor   
88         5.6          3.0         4.1          1.3  Iris-versicolor   
89         5.5          2.5         4.0          1.3  Iris-versicolor   
90         5.5          2.6         4.4          1.2  Iris-versicolor   
91         6.1          3.0         4.6          1.4  Iris-versicolor   
92         5.8          2.6         4.0          1.2  Iris-versicolor   
93         5.0          2.3         3.3          1.0  Iris-versicolor   
94         5.6          2.7         4.2          1.3  Iris-versicolor   
95         5.7          3.0         4.2          1.2  Iris-versicolor   
96         5.7          2.9         4.2          1.3  Iris-versicolor   
97         6.2          2.9         4.3          1.3  Iris-versicolor   
98         5.1          2.5         3.0          1.1  Iris-versicolor   
99         5.7          2.8         4.1          1.3  Iris-versicolor   

    long_petal_minus_media  long_petal_minus_media_square  
50                    0.44                         0.1936  
51                    0.24                         0.0576  
52                    0.64                         0.4096  
53                   -0.26              

As we observed, the calculation was quite overwhelming and that is why the previously exposed methods somehow allow a lot to be achieved in so few lines of code.
For example, we can do the same thing much more efficiently with the `np.std()` method. Taking the column 'long_petalo' as a parameter

In [19]:
devstd_long_petal = np.std(df_iris_versicolor['long_petal'])
print(f'DevSt = {devstd_long_petal}')

DevSt = 0.4651881339845203


# Practice 1: Acquire data from ChEMBL

## Concepts to work
**[Uniprot](https://www.uniprot.org/):** It is a database that seeks to provide the scientific community with a comprehensive, high-quality and freely accessible resource of protein sequences and functional information<sup> **1** </sup>.

**[ChEMBL](https://www.ebi.ac.uk/chembl/):** It is a database that contains bioactive molecules, it gathers chemical, bioactivity and genomic data<sup> **2** </sup>.

**Half-maximal inhibitory concentration (IC50):** Expresses the amount of drug necessary to inhibit a biological process by half the uninhibited value, it is the most widely used measure of drug efficacy or potency<sup> **3** </sup>.

**pIC50:** Is the negative logarithm to the base ten of the IC50, when the units of are **molars (M)**. It is used to facilitate the comparison between different IC50. Also, it is important to know that the higher the pIC50 the drug has a greater efficacy or greater potential<sup> **3** </sup>.

**Half maximal effective concentration (EC50):** It is the effective concentration to produce 50% of the maximum response, it is used to compare the potencies of the drugs. Also, it is important to know that the lower the EC50, the more potent the drug will be. The negative logarithm in base ten **(pEC50)** is also calculated for this request to facilitate its understanding<sup> **4** </sup>.

**Inhibitor constant (Ki):** It is the concentration required to produce half of the maximum inhibition, it is useful to describe the binding affinity of a molecule to a receptor.

**SMILES (Simplified Molecular-Input Line-Entry System):** This is a line notation for describing chemical structures using short ASCII strings<sup> **5** </sup>.

## Problem Statement
For an investigation we want to identify the compounds that act with a specific target, the glycogen synthase kinase-3 beta protein, which acts as a negative regulator in the hormonal control of glucose homeostasis, Wnt signaling and regulation of transcription factors and microtubules .

In this practice we are going to explore and learn about the bioactive compounds of the protein, their structures and some physicochemical characteristics.

For which, we will use the information provided by the **ChEMBL** database, which allows us to filter and download the known bioactivity data of the compounds that interact with our target of interest. Later, we will work the data in a ` DataFrame `, which will allow us to easily organize, visualize and manipulate it.

The first thing to do is connect to **ChEMBL**, using the **webresource client** library.

## Connect to the ChEMBL database

In [20]:
from chembl_webresource_client.new_client import new_client # The webresource client library is imported that allows connecting to ChEMBL
import pandas as pd
import math
from rdkit.Chem import PandasTools

Must be created for API access

In [21]:
targets = new_client.target
compounds = new_client.molecule
bioactivities = new_client.activity

## Target data
Then, we must look for the ID of the target of interest in the Uniprot database, which in this case is Glycogen synthase kinase-3 beta, ID: [P49841](https://www.uniprot.org/uniprot/P49841).

In [22]:
uniprot_id = 'P49841'
# It takes only some information from ChEMBL that is of interest
target_P49841 = targets.get(target_components__accession=uniprot_id) \
                     .only('target_chembl_id', 'organism', 'pref_name', 'target_type')
pd.DataFrame.from_records(target_P49841)

organism                               pref_name target_chembl_id   
0  Homo sapiens         Glycogen synthase kinase-3 beta        CHEMBL262  \
1  Homo sapiens         Glycogen synthase kinase-3 beta        CHEMBL262   
2  Homo sapiens              Glycogen synthase kinase-3    CHEMBL2095188   
3  Homo sapiens  Axin-1/Glycogen synthase kinase-3 beta    CHEMBL3883309   

                   target_type  
0               SINGLE PROTEIN  
1               SINGLE PROTEIN  
2               PROTEIN FAMILY  
3  PROTEIN-PROTEIN INTERACTION

In [23]:
# Select the target of interest
target = target_P49841[0]
print(f'The target of interest is: {str(target)}')

The target of interest is: {'organism': 'Homo sapiens', 'pref_name': 'Glycogen synthase kinase-3 beta', 'target_chembl_id': 'CHEMBL262', 'target_type': 'SINGLE PROTEIN'}


In [24]:
# Save the ChEMBL-ID
chembl_id = target['target_chembl_id']
print(f'The ChEMBL-ID of interest is: {chembl_id}')

The ChEMBL-ID of interest is: CHEMBL262


## Bioactivity data

Now we consult the bioactivity data that is of interest. The steps to follow are:
1. Download and filter bioactivities for the target
     Bioactivity data will be filtered as follows:
         * Type of bioactivity: IC50, EC50, Ki
         * Relationship: "="
2. Convert the downloaded data into a data frame:
     The columns of interest are: `molecule_chembl_id`, `type`, `relation`, `pchembl_value`

In [25]:
# First, we download the entire database
bioact_temp = bioactivities.filter(target_chembl_id = chembl_id)\
                       .filter(relation = '=') \
                       .only('molecule_chembl_id', 'type', 'relation', 'standard_value', 'standard_units', 'pchembl_value', )
df_bioact_temp = pd.DataFrame(bioact_temp)
# rearrange the columns
df_bioact_temp = df_bioact_temp[['molecule_chembl_id', 'type', 'relation', 'value', 'units', 'pchembl_value']]
df_bioact_temp

molecule_chembl_id              type relation  value units pchembl_value
0          CHEMBL322970              IC50        =   0.38    uM          6.42
1          CHEMBL322970              EC50        =   45.6    uM          4.34
2          CHEMBL322970              EC50        =  30.62    uM          4.51
3          CHEMBL322970  Fold stimulation        =    1.3  None          None
4          CHEMBL112564              IC50        =   6.92    uM          5.16
...                 ...               ...      ...    ...   ...           ...
8081      CHEMBL2165191          Delta TM        =   1.33     C          None
8082       CHEMBL215152          Delta TM        =   0.88     C          None
8083      CHEMBL3545096          Delta TM        =   7.52     C          None
8084       CHEMBL513909      % of control        =   69.0     %          None
8085       CHEMBL300138                Kd        =    8.3    nM          8.08

[8086 rows x 6 columns]

In [26]:
# Then, we filter by the type of activity desired
df_bioact = df_bioact_temp[(df_bioact_temp['type'] == 'IC50') |
                              (df_bioact_temp['type'] == 'EC50')|
                              (df_bioact_temp['type'] == 'Ki')]
print(f'Total data loaded: {len(df_bioact)}')

Total data loaded: 3207


In [27]:
# first composites of the dataframe
df_bioact.head()

molecule_chembl_id  type relation  value units pchembl_value
0       CHEMBL322970  IC50        =   0.38    uM          6.42
1       CHEMBL322970  EC50        =   45.6    uM          4.34
2       CHEMBL322970  EC50        =  30.62    uM          4.51
4       CHEMBL112564  IC50        =   6.92    uM          5.16
5       CHEMBL321421  IC50        =   0.41    uM          6.39

Recall that the `.head()` method displays the first five elements of the `dataframe`, however, we can quickly see what elements are in the *realtion* and *type* columns.

In [28]:
df_bioact['relation'].value_counts()

relation
=    3207
Name: count, dtype: int64

In [29]:
df_bioact['type'].value_counts()

type
IC50    2849
Ki       289
EC50      69
Name: count, dtype: int64

Since the *realtion* column has only one type (this is due to the initial database filter), we can remove it:

In [30]:
df_bioact.pop('relation')
df_bioact.head()

molecule_chembl_id  type  value units pchembl_value
0       CHEMBL322970  IC50   0.38    uM          6.42
1       CHEMBL322970  EC50   45.6    uM          4.34
2       CHEMBL322970  EC50  30.62    uM          4.51
4       CHEMBL112564  IC50   6.92    uM          5.16
5       CHEMBL321421  IC50   0.41    uM          6.39

### Clear data
It is possible that some compounds have missing values and also duplicates, since the same compound may have been tested more than once (we will keep only the one that was tested first)

In [31]:
# First we check how many compounds we have in total
ori_len = len(df_bioact)
print(f'Total original compounds is: {ori_len}')

Total original compounds is: 3207


In [32]:
# Remove compounds that do not have pChEMBL_value
df_bioact = df_bioact.dropna(axis=0, how = 'any')
new_len = len(df_bioact)
print(f'Total compounds after removing those with missing data: {new_len}')
# Subtract the total number of compounds from the total number of compounds by removing those without pChEMBL_value
print(f'Total compounds removed {ori_len - new_len}')
ori_len = new_len

Total compounds after removing those with missing data: 3175
Total compounds removed 32


In [33]:
# Duplicate compounds are removed and we are left with the first tested compound
df_bioact = df_bioact.drop_duplicates('molecule_chembl_id', keep = 'first')
new_len = len(df_bioact)
print(f'Total compounds without duplicates : {new_len}')
# Subtract from the total number of compounds by removing those without pChEMBL_value the total number of compounds without duplicates
print(f'Total compounds removed {ori_len - new_len}')
ori_len = new_len

Total compounds without duplicates : 2698
Total compounds removed 477


In [34]:
df_bioact.head()

molecule_chembl_id  type value units pchembl_value
0       CHEMBL322970  IC50  0.38    uM          6.42
4       CHEMBL112564  IC50  6.92    uM          5.16
5       CHEMBL321421  IC50  0.41    uM          6.39
6       CHEMBL388978  IC50  0.16    uM          6.80
8       CHEMBL115875  IC50   9.6    uM          5.02

Now that we have removed some rows we will reset the index so that it is continuous

In [35]:
df_bioact.reset_index(drop=True, inplace=True)
df_bioact.head()

molecule_chembl_id  type value units pchembl_value
0       CHEMBL322970  IC50  0.38    uM          6.42
1       CHEMBL112564  IC50  6.92    uM          5.16
2       CHEMBL321421  IC50  0.41    uM          6.39
3       CHEMBL388978  IC50  0.16    uM          6.80
4       CHEMBL115875  IC50   9.6    uM          5.02

### Organize the data
Let's arrange the DataFrame from largest to smallest pchembl_value. Notice that the column values are not numeric

In [36]:
print(df_bioact['pchembl_value'][0],type(df_bioact['pchembl_value'][0]))

6.42 <class 'str'>


Therefore, we must first convert them to type `float`.

In [37]:
df_bioact['pchembl_value'] = df_bioact['pchembl_value'].astype(float)
print(df_bioact['pchembl_value'][0],type(df_bioact['pchembl_value'][0]))

6.42 <class 'numpy.float64'>


Now we proceed to organize the DataFrame

In [38]:
# Arrange from highest to lowest pchembl_value
df_bioact.sort_values(by="pchembl_value", ascending=False, inplace=True)
# reset the index
df_bioact.reset_index(drop=True, inplace=True)
# We print the first data
df_bioact.head()

molecule_chembl_id  type    value units  pchembl_value
0      CHEMBL2386094  IC50    0.013    nM          10.89
1      CHEMBL3895998  IC50      0.1    nM          10.00
2      CHEMBL4084855  IC50      0.1    nM          10.00
3       CHEMBL564450  IC50  0.00014    uM           9.85
4      CHEMBL3957649  IC50      0.2    nM           9.70

### Save and load the data
To continue using the Data Frame in practice without always having to connect to ChEMBL, we are going to save the obtained Data Frame as a file separated by commas (data/composites_uniprot_id.csv)

In [39]:
df_bioact.to_csv(f"./data/compounds_{uniprot_id}.csv", index=0)

From now on, if we want to use the Dataframe, we can load the saved file

In [40]:
df_bioact = pd.read_csv(f"data/compounds_{uniprot_id}.csv")
df_bioact.head()

molecule_chembl_id  type    value units  pchembl_value
0      CHEMBL2386094  IC50  0.01300    nM          10.89
1      CHEMBL3895998  IC50  0.10000    nM          10.00
2      CHEMBL4084855  IC50  0.10000    nM          10.00
3       CHEMBL564450  IC50  0.00014    uM           9.85
4      CHEMBL3957649  IC50  0.20000    nM           9.70

## Compound data

Next we are going to obtain the data of the molecules that are stored within each molecule_chembl_id

In [41]:
# Load the previously saved file
df_bioact = pd.read_csv(f"data/compounds_{uniprot_id}.csv")

In [42]:
# First we need to get the list of compounds that we define as bioactive
list_comp_id = list(df_bioact['molecule_chembl_id'])
# Get the structure of each compound
compounds_list = compounds.filter(molecule_chembl_id__in = list_comp_id) \
                             .only('molecule_chembl_id','molecule_structures')

In [43]:
# We must convert the obtained list into a dataframe. This may take a few minutes
df_comp = pd.DataFrame(compounds_list)
# remove duplicates
df_comp = df_comp.drop_duplicates('molecule_chembl_id', keep = 'first')
# print(f'The total number of compounds is: ' + str(len(dataframe_comp)))
print(f'Total of compounds is: {str(len(df_comp))}')
df_comp.head()

Total of compounds is: 2698


molecule_chembl_id                                molecule_structures
0         CHEMBL6246  {'canonical_smiles': 'O=c1oc2c(O)c(O)cc3c(=O)o...
1         CHEMBL6291  {'canonical_smiles': 'Cn1cc(C2=C(c3cn(CCCSC(=N...
2       CHEMBL269538  {'canonical_smiles': 'COc1ccc2c(c1)[nH]c1c(C)n...
3           CHEMBL28  {'canonical_smiles': 'O=c1cc(-c2ccc(O)cc2)oc2c...
4         CHEMBL7463  {'canonical_smiles': 'CN(C)CCCn1cc(C2=C(c3c[nH...

Compounds have different types of representations such as the SMILES, the InChI and the InChI Key. We are only interested in staying with the SMILES since it describes the chemical structure.

In [44]:
# We are going to use a for loop to iterate through each row (df_comp.iterrows())
for i, cmpd in df_comp.iterrows():
     if df_comp.loc[i]['molecule_structures'] is not None:
         df_comp.loc[i]['molecule_structures'] = cmpd['molecule_structures']['canonical_smiles']

df_comp.head()

molecule_chembl_id                                molecule_structures
0         CHEMBL6246         O=c1oc2c(O)c(O)cc3c(=O)oc4c(O)c(O)cc1c4c23
1         CHEMBL6291  Cn1cc(C2=C(c3cn(CCCSC(=N)N)c4ccccc34)C(=O)NC2=...
2       CHEMBL269538                      COc1ccc2c(c1)[nH]c1c(C)nccc12
3           CHEMBL28               O=c1cc(-c2ccc(O)cc2)oc2cc(O)cc(O)c12
4         CHEMBL7463  CN(C)CCCn1cc(C2=C(c3c[nH]c4ccccc34)C(=O)NC2=O)...

### Clear salts of smiles
If we look closely at the *smiles* representation of each molecule, we can see that some have *salts* that need to be cleaned up.
First we will filter the molecules that have salts, usually you can see it in the smileies because they have a dot in the text string:

In [45]:
df_comp[df_comp.molecule_structures.str.contains("\.")]

molecule_chembl_id                                molecule_structures
168         CHEMBL69710                                        [Cl-].[Li+]
712        CHEMBL492244  CN1CCN(CCO/N=C2C(=C3/C(=O)Nc4cc(Br)ccc43)/Nc3c...
826        CHEMBL558982     Br.c1ccc(/N=c2/nc(-c3ccccc3)n(-c3ccccc3)s2)cc1
1254      CHEMBL1801613                             Cl.NCC(=O)c1ccc(Br)cc1
1371      CHEMBL2022412  Br.c1ccc(-c2n/c(=N\Cc3cccnc3)sn2-c2cccc3ccccc2...
...                 ...                                                ...
2093      CHEMBL4098458  COc1cc(F)cc(F)c1C1CN(c2nc(-c3ccncc3)cc(=O)n2C)...
2095      CHEMBL4099428  Cl.Cn1c(N2CCNC(c3ccc(Br)cc3)C2)nc(-c2ccncc2)cc1=O
2110      CHEMBL4103742      Cl.Cn1c(N2CCNC(c3ccccc3)C2)nc(-c2ccncc2)cc1=O
2116      CHEMBL4105346  Cl.Cn1c(N2CCN[C@@H](c3ccc(Cl)cc3)C2)nc(-c2ccnc...
2504      CHEMBL4763626  Cl.NCCO/N=C1C(=C2/C(=O)Nc3ccc([N+](=O)[O-])cc3...

[67 rows x 2 columns]

That is, of the 2658 initial compounds, 67 have salts that must be cleaned. We can use a `rdkit` module to clean up salts called `rdkit.Chem.SaltRemover`.
Suppose we want to remove the salt from the smile "CN1CCN(CCO/N=C2C(=C3/C(=O)Nc4cc(Br)ccc43)/Nc3ccccc3/2)CC1.Cl", the following can be done:

In [46]:
# libraries
from rdkit.Chem.SaltRemover import SaltRemover
from rdkit.Chem import MolFromSmiles, MolToSmiles

# The module to remove salts is loaded
remover = SaltRemover()
# convert the smiley to a mol object
mol = MolFromSmiles('Br.CCCCCc1n/c(=N\Cc2cccnc2)sn1-c1ccccc1')
# remove salts (res=molecule, deleted=deleted fragment) smile to a mol object
res, deleted = remover.StripMolWithDeleted(mol)
# convert the mol object to smiles again and remove whitespace
MolToSmiles(res).strip()

'CCCCCc1n/c(=N\\Cc2cccnc2)sn1-c1ccccc1'

Now we are going to create a function that does this process to pass it through the `df_comp` dataframe

In [47]:
def remover_salts(smile):
    remover = SaltRemover()
    mol = MolFromSmiles(smile)
    res, deleted = remover.StripMolWithDeleted(mol)
    return MolToSmiles(res)

We apply the function to the smiles column

In [ ]:
df_comp['molecule_structures'] = df_comp['molecule_structures'].apply(remover_salts)
# Remove whitespace around the edges of each string in the "molecule_structures" column
df_comp["molecule_structures"] = df_comp["molecule_structures"].str.strip()

We look again for smiles that have salts to verify that the cleaning was successful

In [ ]:
df_comp[df_comp.molecule_structures.str.contains("\.")]

We already cleaned the salts from the molecules, now we must remove the molecules with empty data. For example, cleaning the molecule "[Cl-].[Li+]" removes the two salts and leaves the output empty `('')`. To remove these fields we can use `dropna` and filter for those that are not empty.

In [ ]:
df_comp.dropna(subset=["molecule_structures"], inplace=True)
df_comp = df_comp[df_comp['molecule_structures'] != '']

And we already have the set of clean molecules.

## Merge the Dataframes

Now we have two dataframes that we are going to combine to have all the data in a single dataframe and be able to save it.

In [ ]:
# In the Bioactivity Dataframe, only two columns are filtered
df_output = pd.merge(df_bioact[['molecule_chembl_id','pchembl_value']], df_comp, on='molecule_chembl_id')
print(f'Total of compounds is: {str(len(df_output))}')
df_output.head()

Columns can be renamed

In [ ]:
df_output = df_output.rename(columns= {'molecule_structures':'smiles', 'pchembl_value':'pIC50'})
print(f'Total of compounds is: {str(len(df_output))}')
df_output.head()

In order to use the following rdkit function, it is necessary that all the compounds have SMILES, for this reason we eliminate the compounds without SMILES in the dataframe

In [ ]:
df_output = df_output[~df_output['smiles'].isnull()]
print(f'Total of compounds is: {str(len(df_output))}')
df_output.head()

## Draw the molecule

Let's add a new column to the dataframe with the `.AddMoleculeColumnToFrame` function which converts the molecules contained in "smilesCol" into RDKit molecules and adds them to the dataframe

In [ ]:
PandasTools.AddMoleculeColumnToFrame(df_output, smilesCol='smiles')
df_output.head()

Now we can call any compound to see its 2D representation

In [ ]:
df_output.ROMol.iloc[0]

## Save the obtained dataframe

It will save the dataframe as a csv file

In [ ]:
df_output.to_csv(f"data/compounds_{uniprot_id}_full.csv", index=0)

## Practice Activity
Taking into account what you have learned in this Notebook, answer:
1. What other data can we obtain from ChEMBL of the target of interest?
2. Could a new classification of the compounds be carried out filtering them only by the pEC50?, check it and observe the differences
3. Obtain the known data of the compounds, referring to inhibition and activity.

## Conclusions
In this practice, the ChEMBL database was used to obtain data on bioactive compounds against our target of interest. These extracted data in the form of dictionaries and lists were converted into a DataFrame which allows easy visualization of the information obtained. In addition, data from the bioactive compounds were obtained, DataFrames were combined, columns were renamed and a panda tool was used to add a new column to the constructed DataFrame.

# References
1. Coudert, E., Gehant, S., De Castro, E., Pozzato, M., Baratin, D., Neto, T., Sigrist, C. J. A., Redaschi, N., Bridge, A., The UniProt Consortium, Bridge, A. J., Aimo, L., Argoud-Puy, G., Auchincloss, A. H., Axelsen, K. B., Bansal, P., Baratin, D., Neto, T. M. B., Blatter, M.-C., … Wang, Y. (2023). Annotation of biologically relevant ligands in UniProtKB using ChEBI. Bioinformatics, 39(1), btac793. https://doi.org/10.1093/bioinformatics/btac793

2. Mendez, D., Gaulton, A., Bento, A. P., Chambers, J., De Veij, M., Félix, E., Magariños, M. P., Mosquera, J. F., Mutowo, P., Nowotka, M., Gordillo-Marañón, M., Hunter, F., Junco, L., Mugumbate, G., Rodriguez-Lopez, M., Atkinson, F., Bosc, N., Radoux, C. J., Segura-Cabrera, A., … Leach, A. R. (2019). ChEMBL: Towards direct deposition of bioassay data. Nucleic Acids Research, 47(D1), D930-D940. https://doi.org/10.1093/nar/gky1075

3. Aykul, S., & Martinez-Hackert, E. (2016). Determination of half-maximal inhibitory concentration using biosensor-based protein interaction analysis. Analytical Biochemistry, 508, 97-103. https://doi.org/10.1016/j.ab.2016.06.025

4. Waller, D. G., & Sampson, A. P. (2018). Principles of pharmacology and mechanisms of drug action. En Medical Pharmacology and Therapeutics (pp. 3-31). Elsevier. https://doi.org/10.1016/B978-0-7020-7167-6.00001-4

5. Daylight>cheminformatics. (2022). https://www.daylight.com/smiles/
